pythia/utils/text_utils.py

```diff
@@ -271,21 +261,93 @@ def search(self, t, data, scores):
 
         # TODO: Make the data update generic for any type of model
         # This is specific to BUTD model only.
-        data["texts"] = next_word_inds[incomplete_inds].unsqueeze(1)
-        h1 = data["state"]["td_hidden"][0][prev_word_inds[incomplete_inds]]
-        c1 = data["state"]["td_hidden"][1][prev_word_inds[incomplete_inds]]
-        h2 = data["state"]["lm_hidden"][0][prev_word_inds[incomplete_inds]]
-        c2 = data["state"]["lm_hidden"][1][prev_word_inds[incomplete_inds]]
-        data["state"] = {"td_hidden": (h1, c1), "lm_hidden": (h2, c2)}
+        data = update_data(data, prev_word_inds, next_word_inds, incomplete_inds)
 
         next_beam_length = len(prev_word_inds[incomplete_inds])
 
         return False, data, next_beam_length
 
-    def best_score(self):
+    def get_captions(self):
         if len(self.complete_seqs_scores) == 0:
             captions = torch.FloatTensor([0] * 5).unsqueeze(0)
         else:
             i = self.complete_seqs_scores.index(max(self.complete_seqs_scores))
             captions = torch.FloatTensor(self.complete_seqs[i]).unsqueeze(0)
         return captions
+
+class NucleusSampling:
+    def __init__(self, vocab, threshold=0.8):
+        self.vocab = vocab
+        self.vocab_size = vocab.get_size()
+        # Threshold for sum of probability
+        self.threshold = threshold
+        # Lists to store completed sequence
+        self.complete_seq = []
+
+    def init_batch(self, sample_list):
+        setattr(
+            self,
+            "seq",
+            sample_list.answers.new_full(
+                (1, 1), self.vocab.SOS_INDEX, dtype=torch.long
+            ),
+        )
+        # Add a dim and duplicate the tensor beam_size times across that dim
+        sample_list.image_feature_0 = (
+            sample_list.image_feature_0.unsqueeze(1)
+            .expand(-1, 1, -1, -1)
+            .squeeze(0)
+        )
+        return sample_list
+
+    def sample(self, t, data, scores):
+        # Convert scores to probabilities
+        scores = torch.nn.functional.softmax(scores, dim=1)
+        # Sort scores in descending order and then select the top m elements having sum more than threshold.
+        # We get the top_m_scores and their indices top_m_words
+        if t == 0:
```

Comment:
look into if it is possible to converge this into single line instead of if-else. Also try to reuse code from BeamSearch as much as possible after making a base decoder class.

Reply:
This was a required conditional statement. Have reused as much as possible from base decoder class.